Update doc builds to pin the 'pip' version and use tox in readthedocs

sirosen opened the pull request:
No changelog needed. This is a set of minor infrastructure tweaks.

- Pin the version of 'pip' in doc builds to 25.1
- Update readthedocs builds to use tox

##### Contributor checklist

- [x] Included tests for the changes.
- [x] A change note is created in `changelog.d/` (see [`changelog.d/README.md`](https://github.com/jazzband/pip-tools/blob/main/changelog.d/#readme) for instructions) or the PR text says "no changelog needed".

##### Maintainer checklist

- [x] If no changelog is needed, apply the `skip-changelog` label.
- [ ] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

webknjaz commented on .readthedocs.yaml:
```diff
@@ -2,27 +2,23 @@
+    # install uv and use it to get
+    # - Python 3.13
+    # - tox (with tox-uv)
+    - curl -LsSf https://astral.sh/uv/install.sh | sh
```
I'm not happy about the `curl | sh` pattern in here. Can't we just use `asdf` or whatever other dep management tooling is pre-installed in the VMs?

webknjaz replied:
Whatever's fastest I guess..

webknjaz commented on .readthedocs.yaml:
```diff
@@ -2,27 +2,23 @@
+  # control over 'build.commands'
+  # see also: https://github.com/astral-sh/uv/issues/10074#issuecomment-3128225815
+  commands:
+    # a complete clone is needed for setuptools-scm, and this is the method recommended by RTD
+    # ( https://rtd-with-docsearch.readthedocs.io/en/latest/build-customization.html#unshallow-git-clone )
+    - git fetch --unshallow || true
```
This is already included in the tox env. No need to duplicate the effort.
```suggestion
```

webknjaz commented on tox.ini:
Could you also move `git fetch` commands around line 76 into `commands_pre`? I've been doing this everywhere but haven't synced this project, it seems. Although, we could do this separately as it'd affect other docs envs as well...

webknjaz replied:
Marking as resolved.

webknjaz commented on .readthedocs.yaml:
```diff
@@ -2,27 +2,23 @@
-  configuration: docs/conf.py
-  fail_on_warning: true
-
-formats:
-  - pdf
-  - htmlzip
```
So we're loosing these with the update. That's another thing we could mention in the change note.
Alternatively, we could try reproducing the same commands that RTD injects to get these outputs.

webknjaz replied:
I guess it's fine to drop these. Marking the thread as resolved.

webknjaz commented on tox.ini:
```diff
@@ -85,19 +85,19 @@ commands =
-  \t$ python3 -m http.server --directory \
-  \N\{QUOTATION MARK\}\{docs_dir\}\N\{QUOTATION MARK\} 0\n\n" +\
-  "=" * 120)'
+    {posargs:{envdir}/docs_out}
```
I see you didn't migrate the `-t is_unversioned` bit. This is fine for now as it's out of the scope, but would be good to do in a follow-up. It's particularly useful when `setuptools-scm` is used as the version string being dynamic prevents Sphinx from reusing cache and we can optimize it by setting it to something static in local dev envs.

webknjaz commented on docs/requirements.in:
```diff
@@ -1,3 +1,7 @@
+# pip itself is a requirement, so that we keep the version of it pinned
+# for build consistency
+pip
```
Yes, that's my understanding of the root cause as well.

webknjaz replied:
I don't normally do a `skip_install` and try to go for `-c` instead..

sirosen commented:
I've been meaning to get back to some of the other work in pip-tools, but I forgot entirely about this particular PR. I think a rebase and some basic cleanup is something I can do here after my workday. I should re-title the PR as well, seeing as I split #2219 from this.

sirosen commented:
I've come back to this to start trying to clean it up. I retitled, updated the initial doc to indicate that the PR was split up, and removed the skip-changelog labels.

I think we should drop the epub and pdf format docs, which is what's currently done here. They're much less used, and I want the toolchain all aligned so that what we can test locally matches RTD. But that probably needs a changelog note -- not sure why I didn't think so before.

I'm about to start the rebase... :crossed_fingers: that goes smoothly

webknjaz commented on .readthedocs.yaml:
```diff
@@ -2,27 +2,23 @@
 build:
-  os: ubuntu-22.04
-  tools:
-    python: >-
-      3.11
```
Well, this change would need to be integrated in different places. But it's not in the scope of this PR so I'm marking the thread as resolved. This is something to tackle in the future.
